File: graph/BidirectionalBFS/cpp/BidirectionalBFS.cpp
#include "BidirectionalBFS.hpp"

#include <algorithm>
#include <deque>

using namespace std;

namespace {
vector<int> buildPath(int meeting, const vector<int>& parentFromSource, const vector<int>& parentFromTarget) {
    vector<int> left;
    for (int vertex = meeting; vertex != -1;) {
        left.push_back(vertex);
        if (parentFromSource[vertex] == vertex) {
            break;
        }
        vertex = parentFromSource[vertex];
    }
    reverse(left.begin(), left.end());

    vector<int> path = left;
    for (int vertex = parentFromTarget[meeting]; vertex != -1 && vertex != meeting;) {
        path.push_back(vertex);
        if (parentFromTarget[vertex] == vertex) {
            break;
        }
        vertex = parentFromTarget[vertex];
    }
    return path;
}

int expandOneLevel(
    deque<int>& frontier,
    vector<int>& parentThisSide,
    const vector<int>& parentOtherSide,
    const vector<vector<int>>& graph
) {
    const size_t levelSize = frontier.size();
    for (size_t i = 0; i < levelSize; ++i) {
        const int current = frontier.front();
        frontier.pop_front();

        for (const int neighbor : graph[current]) {
            if (parentThisSide[neighbor] != -1) {
                continue;
            }

            parentThisSide[neighbor] = current;
            if (parentOtherSide[neighbor] != -1) {
                return neighbor;
            }
            frontier.push_back(neighbor);
        }
    }
    return -1;
}
}  // namespace

BidirectionalBFSResult shortestPathBidirectionalBFS(
    const vector<vector<int>>& graph,
    int source,
    int target
) {
    if (source == target) {
        return {0, {source}};
    }

    const int n = static_cast<int>(graph.size());
    vector<int> parentFromSource(n, -1);
    vector<int> parentFromTarget(n, -1);
    deque<int> sourceFrontier;
    deque<int> targetFrontier;

    parentFromSource[source] = source;
    parentFromTarget[target] = target;
    sourceFrontier.push_back(source);
    targetFrontier.push_back(target);

    while (!sourceFrontier.empty() && !targetFrontier.empty()) {
        int meeting = -1;
        if (sourceFrontier.size() <= targetFrontier.size()) {
            meeting = expandOneLevel(sourceFrontier, parentFromSource, parentFromTarget, graph);
        } else {
            meeting = expandOneLevel(targetFrontier, parentFromTarget, parentFromSource, graph);
        }

        if (meeting != -1) {
            vector<int> path = buildPath(meeting, parentFromSource, parentFromTarget);
            return {static_cast<int>(path.size()) - 1, path};
        }
    }

    return {-1, {}};
}

File: graph/BidirectionalBFS/rust/src/bidirectional_bfs.rs
use std::collections::VecDeque;

pub struct BidirectionalBFSResult {
    pub distance: isize,
    pub path: Vec<usize>,
}

pub fn shortest_path_bidirectional_bfs(
    graph: &[Vec<usize>],
    source: usize,
    target: usize,
) -> BidirectionalBFSResult {
    if source == target {
        return BidirectionalBFSResult {
            distance: 0,
            path: vec![source],
        };
    }

    let n = graph.len();
    let mut parent_from_source = vec![None; n];
    let mut parent_from_target = vec![None; n];
    let mut source_frontier = VecDeque::new();
    let mut target_frontier = VecDeque::new();

    parent_from_source[source] = Some(source);
    parent_from_target[target] = Some(target);
    source_frontier.push_back(source);
    target_frontier.push_back(target);

    while !source_frontier.is_empty() && !target_frontier.is_empty() {
        let meeting = if source_frontier.len() <= target_frontier.len() {
            expand_one_level(
                &mut source_frontier,
                &mut parent_from_source,
                &parent_from_target,
                graph,
            )
        } else {
            expand_one_level(
                &mut target_frontier,
                &mut parent_from_target,
                &parent_from_source,
                graph,
            )
        };

        if let Some(meeting_vertex) = meeting {
            let path = build_path(meeting_vertex, &parent_from_source, &parent_from_target);
            return BidirectionalBFSResult {
                distance: path.len() as isize - 1,
                path,
            };
        }
    }

    BidirectionalBFSResult {
        distance: -1,
        path: Vec::new(),
    }
}

fn expand_one_level(
    frontier: &mut VecDeque<usize>,
    parent_this_side: &mut [Option<usize>],
    parent_other_side: &[Option<usize>],
    graph: &[Vec<usize>],
) -> Option<usize> {
    let level_size = frontier.len();
    for _ in 0..level_size {
        let current = frontier.pop_front().expect("frontier size was measured");
        for &neighbor in &graph[current] {
            if parent_this_side[neighbor].is_some() {
                continue;
            }

            parent_this_side[neighbor] = Some(current);
            if parent_other_side[neighbor].is_some() {
                return Some(neighbor);
            }
            frontier.push_back(neighbor);
        }
    }
    None
}

fn build_path(
    meeting: usize,
    parent_from_source: &[Option<usize>],
    parent_from_target: &[Option<usize>],
) -> Vec<usize> {
    let mut left = Vec::new();
    let mut vertex = Some(meeting);
    while let Some(current) = vertex {
        left.push(current);
        vertex = parent_from_source[current];
        if vertex == Some(current) {
            vertex = None;
        }
    }
    left.reverse();

    let mut path = left;
    vertex = parent_from_target[meeting];
    while let Some(current) = vertex {
        if current == meeting {
            break;
        }
        path.push(current);
        vertex = parent_from_target[current];
        if vertex == Some(current) {
            vertex = None;
        }
    }
    path
}
